### PR TITLE
General improvements to hearting and queuing

### LINKED ISSUE #398: Implement hearting developer levels
hearting developer levels doesn't work as /favourite/slot/developer/{id} isn't implemented

as a side note neither is /slots/developer, which just returns active player counts for developer levels

### LINKED ISSUE #409: Hearted levels are in incorrect order
Hearted levels are supposed to be listed from latest to oldest, but right now they're listed from oldest to latest

this causes the wrong level to be shown as your latest hearted in LBP2.

### PR DESCRIPTION
**Pull request by TorutheRedFox**
Fixes #398
Fixes #409 

In detail, what this does is it allows developer levels to be hearted, sorting the slots to their particular game (LBP3 was disregarded as you can't heart developer levels at all), which allows them to be sorted using LBP2's game filters, which also got implemented for hearted and queued lists

hearted and queued levels are also now sorted from latest to oldest, which makes the latest hearted level in LBP2 and LBP3 show up correctly (previously it showed the first level that you hearted)

### CODE REVIEW

**Comment by TorutheRedFox** [comment]
I have no idea how LBP1 requests hearted levels

**Comment by TorutheRedFox** [comment]
I need a packet capture for this stuff lol (but I can't find any except for a capture for vita from someone who hasn't hearted literally anything)

**Comment by TorutheRedFox** [comment]
the server isn't even sending the slots across and idk why
great

**Comment by TorutheRedFox** [comment]
like, is the query messed up or what?

**Comment by TorutheRedFox** [comment]
ok so

-the slots are in the database
-the hearted slot list does have the developer slot in it
-the slot doesn't seem to be picked up by the query

**Comment by Slendy** [comment]
It's because the current implementation uses ByGameVersion which adds a query so that only user slots are returned.

**Comment by TorutheRedFox** [comment]
oh

well time to implement bygameversion for favourite levels lol

**Comment by TorutheRedFox** [comment]
wow I can't believe I missed that lmao

I think I'm done here

**Comment by TorutheRedFox** [comment]
BCES00141_screenshot_2022_08_06_22_02_11

**Comment by TorutheRedFox** [comment]
wait hold on, lbp2 levels still are broken

**Comment by TorutheRedFox** [comment]
fixed it

**Comment by TorutheRedFox** [comment]
the fix is pretty hacky but it gets the job done

**Comment by TorutheRedFox** [comment]
actually thinking about it this probably would need a way to deal with dlc levels, but I don't know whether there's a way to actually distinguish between an LBP1 level or an LBP2 level

**Comment by TorutheRedFox** [comment]
removed the sorting

the game can probably handle it itself

**Comment by TorutheRedFox** [comment]
fixed #409 while I was at it

**Comment by TorutheRedFox** [comment]
ok upon further research, the janky way of setting the level could theoretically work as you seemingly can't heart DLC levels in LBP1, but it'd cause issues with people modding LBP2 to get LBP1's story slots, so a hardcoded list of LBP1 story slots could theoretically work?

**Comment by TorutheRedFox** [comment]
ok so what I'm doing right now is testing the hardcoded slot guids for LBP1 story levels, and better sorting of lolcatftw and favouriteSlots (with filter support)

**Comment by TorutheRedFox** [comment]
if anyone wants to find the first slot guid for LBP2 you can go ahead

**Comment by TorutheRedFox** [comment]
it seems to be g124806

**Comment by TorutheRedFox** [comment]
ok it seems to work correctly

**Comment by TorutheRedFox** [comment]
at this point might aswell call this general improvements as I've not only fixed up hearting developer levels but also the way hearted and queued levels in general are listed

**Comment by TorutheRedFox** [comment]
damn I'm on a roll lol

I'm in the middle of implementing text filter for everything that the game supports it for lol

**Comment by TorutheRedFox** [comment]
if anyone wants to know how my textfiltering implementation looks like, take a look at it on the main branch of my fork (https://github.com/TorutheRedFox/ProjectLighthouse/tree/main)

**Comment by TorutheRedFox** [comment]
had to undeprecate HeartedProfileId as it's actually needed for proper sorting

**Comment by TorutheRedFox** [comment]
hold on gotta rework some of this

**Comment by TorutheRedFox** [comment]
fixed #430

File: ProjectLighthouse/PlayerData/Profiles/HeartedProfile.cs [hunk 1]
```diff
@@ -6,9 +6,7 @@ namespace LBPUnion.ProjectLighthouse.PlayerData.Profiles;
 
 public class HeartedProfile
 {
-    // ReSharper disable once UnusedMember.Global
-    [Obsolete($"Use {nameof(HeartedUserId)} instead, this is a key which you should never need to use.")]
-    [Key]
+    [Key] // this is actually useful for sorting
```
Comment (jvyden) [inline ProjectLighthouse/PlayerData/Profiles/HeartedProfile.cs:9]:
```suggestion
    [Key]
```

File: ProjectLighthouse.Servers.GameServer/Controllers/Slots/ListController.cs [hunk 2]
```diff
@@ -131,29 +142,45 @@ public async Task<IActionResult> GetFavouriteSlots(string username, [FromQuery]
...
+        if (SlotHelper.IsTypeInvalid(slotType)) return this.BadRequest();
+
+        if (slotType == "developer") id = await SlotHelper.GetPlaceholderSlotId(this.database, id, SlotType.Developer);
+
         Slot? slot = await this.database.Slots.FirstOrDefaultAsync(s => s.SlotId == id);
         if (slot == null) return this.NotFound();
 
+        // assume g124806 is first LBP2 slot as it's the first guid in lbp2story_slots.slt, with guids being shared in the cutdown lbp2beta_slots.slt
+        PlayerData.GameVersion slotGameVersion = (slot.InternalSlotId < 124806) ? PlayerData.GameVersion.LittleBigPlanet1 : token.GameVersion;
+        if (slotType == "developer") slot.GameVersion = slotGameVersion;
```
[the remark this answers is not in the document]
Comment (jvyden) [reply ProjectLighthouse.Servers.GameServer/Controllers/Slots/ListController.cs:182]: Also, 124806 should be a constant, `FirstLbp2DeveloperSlotId` or something.

File: ProjectLighthouse.Servers.GameServer/Controllers/Slots/ListController.cs [hunk 3]
```diff
@@ -131,29 +142,45 @@ public async Task<IActionResult> GetFavouriteSlots(string username, [FromQuery]
...
```
Comment (jvyden) [inline ProjectLighthouse.Servers.GameServer/Controllers/Slots/ListController.cs:160]: Same as my other thread

**Review by jvyden — CHANGES_REQUESTED** [review]
Good work! There's a few issues I see mostly relating to duplicated code, but other than that it's just nitpicks.

File: ProjectLighthouse.Servers.GameServer/Controllers/Slots/ListController.cs [hunk 4]
```diff
@@ -131,29 +142,45 @@ public async Task<IActionResult> GetFavouriteSlots(string username, [FromQuery]
...
 
+        if (SlotHelper.IsTypeInvalid(slotType)) return this.BadRequest();
+
+        if (slotType == "developer") id = await SlotHelper.GetPlaceholderSlotId(this.database, id, SlotType.Developer);
+
         Slot? slot = await this.database.Slots.FirstOrDefaultAsync(s => s.SlotId == id);
         if (slot == null) return this.NotFound();
 
+        // assume g124806 is first LBP2 slot as it's the first guid in lbp2story_slots.slt, with guids being shared in the cutdown lbp2beta_slots.slt
+        PlayerData.GameVersion slotGameVersion = (slot.InternalSlotId < 124806) ? PlayerData.GameVersion.LittleBigPlanet1 : token.GameVersion;
```
[the remark this answers is not in the document]
Comment (TorutheRedFox) [reply ProjectLighthouse.Servers.GameServer/Controllers/Slots/ListController.cs:181]: that's some weird thing Visual Studio did that I didn't notice

File: ProjectLighthouse/Extensions/DatabaseExtensions.cs [hunk 5]
```diff
@@ -39,6 +39,31 @@ public static IQueryable<Slot> ByGameVersion
...
+        {
+            // prevent developer slots from showing up in a newer game
+            query = query.Where(s => (s.Type == SlotType.User && s.GameVersion <= gameVersion) || (s.Type == SlotType.Developer && s.GameVersion == gameVersion));
+        }
+
+        if (!includeSublevels) query = query.Where(s => !s.SubLevel);
+
+        return query;
+    }
+
```
[the remark this answers is not in the document]
Comment (TorutheRedFox) [reply ProjectLighthouse/Extensions/DatabaseExtensions.cs:66]: it also selects developer slots as ByGameVersion specifically excludes them

[the remark this answers is not in the document]
Comment (TorutheRedFox) [reply ProjectLighthouse/Extensions/DatabaseExtensions.cs:66]: thinking about it I can reduce it using an optional parameter

[the remark this answers is not in the document]
Comment (TorutheRedFox) [reply ProjectLighthouse/Extensions/DatabaseExtensions.cs:66]: ah looking back I did that already but forgot to remove the old function

File: ProjectLighthouse.Servers.GameServer/Controllers/Slots/ListController.cs [hunk 6]
```diff
@@ -227,4 +255,87 @@ public async Task<IActionResult> RemoveFavouriteUser(string username)
...
+            "lbp1" => GameVersion.LittleBigPlanet1,
+            "lbp2" => GameVersion.LittleBigPlanet2,
+            "lbp3" => GameVersion.LittleBigPlanet3,
+            "both" => GameVersion.LittleBigPlanet2, // LBP2 default option
+            null => GameVersion.LittleBigPlanet1,
+            _ => GameVersion.Unknown,
+        };
+    }
+
+    private IQueryable<Slot> filterLolcatftwByRequest(string? gameFilterType, string? dateFilterType, GameVersion version, string username)
```
[the remark this answers is not in the document]
Comment (TorutheRedFox) [reply ProjectLighthouse.Servers.GameServer/Controllers/Slots/ListController.cs]: while reducing the functions into one this became unnecessary

**Comment by TorutheRedFox** [comment]
ok I think it should be fine now

**Review by jvyden — COMMENTED** [review]
Code looks fine to me now; going to test in a bit and we'll go from there

**Review by jvyden — APPROVED** [review]
LGTM, thank you for your contribution!

**Pull request merged**